README
- Intro:

This is Bookboot, a facility for putting together an OS.IMG file 
 for the Psion Netbook and similar machines. It was developed on a 
 malaybook (Malaysian variety of the netbook), and has been reported to 
 work on standard netbooks and 7books (Series 7 with netbook personality
 module).
 The file format was put together by guesswork and trial and error.

- creating OS.IMG:
  - (optional) adjust parameters in cfg.pl
  - move linux kernel file to 'zImage'
  - (optional) move compressed initrd file to 'initrd.gz'
  - make
  
- Usage:
 - backup any relevant data from the netbook to a safe place (NB:
 	all memory contents and settings will be lost upon the hard
	reset below).
 - backup your original OS.IMG file	
 - copy OS.IMG to dos-formatted first partition of CF card.
 - (for debugging output: attach to serial port at 115kbit 8N1)
 - insert CF card in Netbook
 - do a full reboot (take out all batteries and power supplies, wait
 	a few seconds, reinsert power)
 - watch the debugging output on the serial port,
 	or see the kernel boot (at the time of writing, the lcd is barely
	functional; the red led should flash to indicate that the kernel is
	running).
 - enjoy!

- Internals:
  The relevant part of the OS.IMG (from the perspective of the builtin 
  Malaybook bootloader) appears to be 
   
   wrapper (typically 256 bytes):
   - 24 bytes of label
   - 1 word length of file (or somewhat larger than that)
   - 1 word length of wrapper (only tested with 0x100, i.e. 256 bytes)
   - filler for rest of wrapper
   
   header (typically 256 bytes):
   - first word: not used?
   - second word: entry point (typically jump to start of bootcode)
   - at offset 0x90: file size	

   The rest is up to you, typically:
   - filler to end of header 
   - bootcode
   - your choice of code..
   rest:
   - up to you...

   the only 'magic' involved seems to be that _if_ the os.img file is
   larger than 16kB+256 bytes, the bootloader code checks for 16 magic
   bytes at locations 0x41d4-0x41e3 in the file.
   
   The Malaybook bootloader will discard the wrapper, load the rest of the
   file at memory location 0xc8000000, and then jump to location
   0xc8000004 (i.e., the second word of the header).

   The file layout used by bookboot is
   - 0x0   256 byte wrapper
   - 0x100 256 byte header, starts with two jump instructions to bootcode below
   - 0x200 256 bytes of parameters for bootcode
   - 0x300 bit of bootcode 
   - 0x2100: list of tagged entries containing boot parameters for linux kernel
   - 0x41d4-0x41e3: 16 magic bytes
   - 0x8100: linux zImage
   - a bit of spacer (apparently necessary to avoid overwriting the initrd
   	during kernel decompression)
   - (optional) initrd.gz, compressed initial ramdisk

   The bootcode echos some info to the serial port, then transfers 
   the taglist to 0xc01e8000 (where the netbook kernel expects it), 
   and finally jumps to the kernel decompression code at 0xc8008000

- Code/package commentary:
   - the package provides the bootcode in both source and binary form.
     The rationale of including the binary is that people who want to use
     bookboot will not need a multi-megabyte [cross-]compiler for what amounts
     to less than 150 lines of arm assembly.
   - For the same reason, the assembly code is designed to be as 'dumb'
     as possible. The only exception is memory autodetection, since this
     can only be done at runtime.
     All parameters are set by a perl script. Thus, anyone with
     a working perl implementation (and some knowledge of 'make') should be
     able to create an os.img file using this package. 
     
   - For those looking for more options and/or possibilities: check out
     the source, it should be reasonably self-explanatory.

Klaasjan van Druten, July 2004   


Changelog
v0.25:  include automatic memory detection
	enable config file

v0.20: initial release, July 2, 2004



cfg.pl
# cfg.pl
# configuration file for bookboot
# this is read by the perl script bookglue.pl
# 
# change the variables below to your liking
# 


# linux kernel commandline

$cmdline = 'video=sa1100 root=/dev/ram0  rw';


# 0 for silent operation, 1 for some more explanation

$verbose = 0;  


# possible values: "auto", "set", "none"
# "auto"  : do memory detection at boot time
# "set"   : set memory config in taglist according to 
# 	     $memsize
# "none"  : none of the above, i.e. no  explicit memory 
# 	    configuration

$memdetect = "auto"; 


# memory size in MB; possible values: 32, 48 and 64
# only relevant if $memdetect equals "set"

$memsize = 32; 


# filenames to be used
# NB: changing these is ok, but 'make' may be confused;
# you may need to run bookglue.pl manually 
$kernel = "zImage";
$initrd = "initrd.gz";
$image = "os.img";



serial.S
/* 
 * serial.S
 * 
 * modified (mainly simplified) version of:
 *   linux/arch/arm/kernel/debug-armv.S  (from linux-2.4.21-rmk2)
 *   which is Copyright (C) 1994-1999 Russell King
 * 
 * modifications Copyright (C) 2004 Klaasjan van Druten
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 */

#define UTCR3           0x0c
#define UTCR3_RXE       0x00000001      /* Receive Enable                  */
#define UTCR3_TXE       0x00000002      /* Transmit Enable                 */

#define UTSR1           0x20
#define UTSR1_TNF       0x00000004      /* Transmit FIFO Not Full (read)   */
#define UTSR1_TBY       0x00000001      /* Transmitter BusY (read)         */

#define UTDR            0x14

/* Find address of first active UART (serial I/O)
 * note: we assume the MMU is off (otherwise we would need to know where
 * the bootloader maps the serial port).
 * clobbers r1
 */
		.macro	addruart,rx
		mov	\rx, #0x80000000	@ physical base address
		add	\rx, \rx, #0x00050000

		@ see if Ser3 is active
		ldr	r1, [\rx, #UTCR3]
		tst	r1, #UTCR3_TXE

		@ if Ser3 is inactive, then try Ser1
		addeq	\rx, \rx, #(0x00010000 - 0x00050000)
		ldreq	r1, [\rx, #UTCR3]
		tsteq	r1, #UTCR3_TXE

		@ if Ser1 is inactive, then try Ser2
		addeq	\rx, \rx, #(0x00030000 - 0x00010000)
		ldreq	r1, [\rx, #UTCR3]
		tsteq	r1, #UTCR3_TXE

		@ if all ports are inactive, then there is nothing we can do
		moveq	pc, lr
		.endm

		.macro	senduart,rd,rx
		str	\rd, [\rx, #UTDR]
		.endm

		.macro	waituart,rd,rx
1001:		ldr	\rd, [\rx, #UTSR1]
		tst	\rd, #UTSR1_TNF
		beq	1001b
		.endm

		.macro	busyuart,rd,rx
1001:		ldr	\rd, [\rx, #UTSR1]
		tst	\rd, #UTSR1_TBY
		bne	1001b
		.endm

/*
 * Useful debugging routines, argument in r0 [+r1], clobber r1-r3
 */
printhex8:
		mov	r1, #8
		b	printhex

printhex4:
		mov	r1, #4
		b	printhex

printhex2:
		mov	r1, #2
printhex:	adr	r2, hexbuf
		add	r3, r2, r1
		mov	r1, #0
		strb	r1, [r3]
1:		and	r1, r0, #15
		mov	r0, r0, lsr #4
		cmp	r1, #10
		addlt	r1, r1, #'0'
		addge	r1, r1, #'a' - 10
		strb	r1, [r3, #-1]!
		teq	r3, r2
		bne	1b
		mov	r0, r2
		b	printascii

printascii:
		addruart r3
		b	2f
1:		waituart r2, r3
		senduart r1, r3
		busyuart r2, r3
		teq	r1, #'\n'
		moveq	r1, #'\r'
		beq	1b
2:		teq	r0, #0
		ldrneb	r1, [r0], #1
		teqne	r1, #0
		bne	1b
		mov	pc, lr

printch:
		addruart r3
		mov	r1, r0
		mov	r0, #0
		b	1b

hexbuf:		.space 16



bookboot.S
/*
 * bookboot.S
 * part of the bookboot package
 * 
 *  Copyright (C) 2003-2004 Klaasjan van Druten
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 */

/* parameter locations with respect to _start:
 * these should be consistent with bookglue.pl
 */
#define CFG_POS  256
#define CFG_PARS 260
#define CPY_PARS 264
#define REG_PARS 436
#define WELCOME  480

/*
 * possibly detect memory layout and add to taglist
 * copy memory blocks here and there, load registers, and jump somewhere
 * output to serial port
 *
 * register usage:
 * r0    : local/scratch
 * r1-r3 : clobbered by serial routines
 * r4-r6 : memcpy registers
 * r7    : index in parameters
 */
	.align
	.text
	.global _start
_start:	
 	b	realstart
	b	realstart
	.rept 126	@ fill header to 512 bytes
	.long   0x0
	.endr
realstart:
	adr	r0, _start + WELCOME
	bl	printascii
	mov  	r0, #'p'
	bl 	printch
	mov  	r0, pc
	bl 	printhex8
	mov	r0, #'\n'
	bl 	printch
	adr 	r7, _start + CFG_POS
	ldmia	r7, {r4}
	tst	r4, #1
	beq	noauto

	/* memory autodetection */
	/* register usage: r4 taglist index 
	 * r8 memory index
	 * r5,r6,r7, r8: ATAG_MEM
	 * r0-r3: scratch + serial output */
	adr	r7, _start + CFG_PARS
	ldmia	r7, {r4}
3:
	mov	r0, #'t'	@ walk through taglist, use r4 as index
	bl	printch
	mov	r0, r4
	bl	printhex8
	mov	r0, #'\n'
	bl 	printch
	ldmia	r4, {r5, r6}	
	add	r4, r4, r5, lsl #2
	teq	r5, #0
	bne	3b 
	mov	r5, #0x4	@ set up ATAG_MEM
	mov	r6, #0x54000000
	add	r6, r6, #0x00410000
	add	r6, r6, #0x00000002
	mov	r7, #0x01000000 @ 16 MB block size
	mov	r8, #0xc0000000

2:	mov	r0, #'m'	@ test memory presence by verified writing
	bl	printch
	mov	r0, r8
	bl	printhex8
	ldmia	r8, {r0, r1}
	stmia	r8, {r5, r6}
	ldmia	r8, {r2, r3}
	teq	r2, r5
	teqeq	r3, r6
	movne	r0, #'n'
	bne	1f
	stmia	r8, {r0, r1}  	@ restore
	stmia	r4!,{r5-r8}	@ add to taglist
	mov	r0, #'y'
1:	bl	printch
	mov	r0, #'\n'
	bl	printch
	teq	r8, #0xd8000000
	addne	r8, r8, #0x08000000
	bne	2b
	mov	r0, #0
	mov	r1, r0
	stmia	r4!, {r0,r1}
	
noauto:
	adr 	r7, _start + CPY_PARS
2:	ldmia   r7!, {r4, r5, r6}  
	teq	r4, r5
	beq 	exit
	mov	r0,r4      /* print source -> destination */
	bl 	printhex8  
	mov 	r0, #'-'
	bl 	printch
	mov 	r0, #'>'
	bl 	printch
	mov	r0,r6
	bl 	printhex8
	mov 	r0, #'\n'
	bl 	printch

1:	ldr 	r0, [r4], #4	/* copy a word */
	str 	r0, [r6], #4
	subs	r0, r5, r4		/* at end? */
	bne 	1b
	mov	r0, r4		/* print final source location */
	bl 	printhex8
	mov 	r0, #'\n'
	bl 	printch
	b	2b

exit:
	adr 	r7, _start + REG_PARS
	ldmia 	r7, {r0-r6,r8}
	mov 	pc, r8
	

#include "serial.S" 

end:
	.end
	


bookglue.pl
#!/usr/bin/perl
use strict;
#
# Script to create a bootable OS.IMG image for Malaysian variety of Psion Netbook
# Copyright (C) 2003-2004 Klaasjan van Druten 
# 
# based on:
# proglue.pl (part of proboot, v0.2.2)
# 	Script to create a bootable sys$rom.bin image for Psion 5mx Pro
# 	Copyright 2002 Tony Lindgren 
#

my $version = "Bookboot v 0.25\n";  # max 31 characters!

my $cfgfile = "cfg.pl";

# these should match with bookboot.S
my $cfg_pos = 256;
my $cfg_pars = 260;
my $cpy_pars = 264;
my $reg_pars = 436;
my $welcome = 480;

# default values, can be changed in cfgfile
my $memdetect = "auto"; 
my $memsize = 32; 
my $cmdline = 'video=sa1100  root=/dev/ram0  rw';
my $verbose = 0; 
my $kernel = "zImage";
my $initrd = "initrd.gz";
my $image = "os.img";

# size of header/wrapper used by the Netbook boot loader 
my $wrappersize = 0x0100;

my $loadadr = 0xc8000000;
my $kernelpos = 0x08000;
my $taglistpos = 0x02000;
my $taglistdest = 0xc01e8000;
my $initrdspacer=0x10000;
		
my $bootcode = "bookboot.bin";

eval `cat $cfgfile` or die "couldn't run configuration file $cfgfile ";
my $cfg = ( 1 & ($memdetect eq "auto"));
my $kernelsize = -s $kernel;
my $initrdsize = -s $initrd;
my $cmdlinesize = length($cmdline) +1;
if ($cmdlinesize % 4) {
	$cmdlinesize = $cmdlinesize + 4 - ($cmdlinesize % 4);
}
my $var;

my $initrdpos= $kernelpos+$kernelsize+$initrdspacer;
$initrdpos=  $initrdpos - ( $initrdpos % 4096 ) + 4096;

my $sloppysize= $wrappersize+$initrdpos+$initrdsize;
$sloppysize=  $sloppysize - ( $sloppysize % 4096 ) + 4096;

verbprint($version);
verbprint("Creating image file $image\n");
pad_file($image, $wrappersize+$kernelpos, 1);

verbprint("Injecting key\n");
my ($keypos,$key) = malaykey();
inject_to_file($image,$key, $keypos);


if ($kernelsize > 0){
	verbprint("Appending kernel file $kernel\n");
	concat_files($image, $kernel);
}

if ($initrdsize > 0){
	verbprint("Appending initrd file $initrd\n");
	pad_file($image, $initrdpos + $wrappersize - ( -s $image));
	concat_files($image, $initrd);
	pad_file($image, $sloppysize  - ( -s $image));
}

verbprint("Injecting boot code file $bootcode\n");
$var = `cat $bootcode`;
inject_to_file($image,  $var, $wrappersize);
$var =  pack('L',$loadadr+$taglistpos); # taglist start
$var .= pack('L',$loadadr+$taglistpos+ 0x2000); # taglist end
$var .= pack('L',$taglistdest);	#   taglist destination
inject_to_file($image, $var, $wrappersize+$cpy_pars);
inject_to_file($image, pack('L',$cfg), $wrappersize+$cfg_pos);
if ($memdetect eq "auto") {
	inject_to_file($image, pack('L',$loadadr+$taglistpos), 
		$wrappersize+$cfg_pars);
}	

$var = pack('L',0x0);   	# r0 = 0
$var .= pack('L',0x40);		# r1 = arch_psion_series7
$var .= pack('L',$taglistdest);	# r2 = taglist position, cf. arlo
$var .= pack('L',0x0);   	# r3-r6 
$var .= pack('L',0x0);   	# r3-r6 
$var .= pack('L',0x0);   	# r3-r6 
$var .= pack('L',0x0);   	# r3-r6 
$var .= pack('L',$loadadr+$kernelpos);   	# jump to kernel
inject_to_file($image, $var, $wrappersize+$reg_pars);
inject_to_file($image, $version, $wrappersize+$welcome);


verbprint("Injecting file size info\n");
inject_to_file($image, pack('L',$sloppysize), 24);
inject_to_file($image, pack('L',$wrappersize), 28);
inject_to_file($image, pack('L', $sloppysize), $wrappersize + 0x90);

verbprint("Injecting linux taglist\n");
my $taglist= build_taglist();
inject_to_file($image, $taglist, $wrappersize+$taglistpos);

verbprint("Done.\n");

# subroutines

sub malaykey(){
	my $keypos= 0x41d4;
	my $key='';
	$key .= pack('L',0xf07c2691);
	$key .= pack('L',0x3157eab3);
	$key .= pack('L',0xd04e82df);
	$key .= pack('L',0xf8010000);
	return ($keypos,$key);
}	
# 
sub build_taglist(){
	my $taglist;
	$taglist  = pack('L', 0x02);        # tag size
	$taglist .= pack('L', 0x54410001);  # ATAG_CORE

	if ($memdetect eq "set"){
		$taglist .= atag_mem(0xc0000000);
		$taglist .= atag_mem(0xc8000000);
		if ($memsize > 32){ $taglist .= atag_mem(0xd0000000) };
		if ($memsize > 48){ $taglist .= atag_mem(0xd8000000) };
	}

	$taglist .= pack('L', $cmdlinesize /4 +2);  # tag size
	$taglist .= pack('L',0x54410009);  	  # ATAG_CMDLINE
	$taglist .= pack("Z$cmdlinesize",$cmdline);

	if ( $initrdsize > 0) {
		$taglist .= pack('L', 0x04);  # tag size
		$taglist .= pack('L', 0x54420005);  # ATAG_INITRD2
		$taglist .= pack('L', $initrdpos+$loadadr);  # start
		$taglist .= pack('L', $initrdsize);  # size
	} 

	$taglist .= pack('L',0x0);  # tag size
	$taglist .= pack('L',0x0);  # ATAG_NONE

	return $taglist;
}

sub atag_mem(){
	my ($loc) = @_;
	my $tag;
	$tag  = pack('L', 0x04);        # tag size
	$tag .= pack('L', 0x54410002);  # ATAG_MEM
	$tag .= pack('L', 0x01000000);  #  16MB block
	$tag .= pack('L', $loc);        # memstart
	return $tag;
}

sub inject_to_file(){
	my ($file,$var,$seek) = @_;
	open(OUT, "+<$file") or die "can't update $file: $!";
	seek(OUT, $seek, 0);
	print OUT $var;
	close OUT;
	verbprint("\tinjected ".length($var)." bytes in $file at location $seek\n");
}

sub append_to_file(){
	my ($file,$var) = @_;
	open(OUT, ">> $file") || die "can't append to $file: $!";
	print OUT $var;
	close OUT;
	verbprint("\tappended ".length($var)." bytes to $file\n");
}


sub concat_files() {
	my($file1, $file2) = @_;
	open OUT, ">>$file1"
		or die "ERROR: Cannot open file ".$file1;
	open IN, "<$file2"
		or die "ERROR: Cannot open file ".$file2;
	while (<IN>) {
		print OUT;
	}
	close (IN);
	close (OUT);
	verbprint("\tappended $file2 to $file1\n");
}

sub pad_file() {
	my($file, $padsize, $overwrite) = @_;
	if ($overwrite gt 0) {
		open OUT, ">$file" or die "ERROR: Cannot open file ".$file;
		verbprint("\tcreating file $file with $padsize bytes\n");
	} else {
		open OUT, ">>$file"  or die "ERROR: Cannot open file ".$file;
		verbprint("\tpadding file $file with $padsize bytes\n");
	}
	for (my $i=0; $i < $padsize; $i++) {
		print OUT "\0";
	}
	close (OUT);
	verbprint("\tsize of $file now ".(-s $file)." bytes\n");
}

sub verbprint() {
        if ($verbose) { print(@_) };
}

